File: containers/scraper/python/scraper.py
import aiohttp
import asyncio
import os
from time import perf_counter
from typing import List, Dict
from redis.asyncio import Redis as AsyncRedis
from itertools import product
import json
from ping3 import ping

async def write_to_redis(redis: AsyncRedis, metric: str, host: str, container_name: str, value: str, window_size: int) -> None:
    
    # Remove port from host
    if ":" in host:
        host = host.split(":")[0]
    
    if container_name != None:
        # Get key
        key = f"metric:{metric}:host:{host}:container:{container_name}"
        
        # Store the metric in Redis
        await redis.rpush(key, value)
        
        # Keep only the recent values
        if await redis.llen(key) > window_size:
            await redis.lpop(key)
    else:
        # Get key
        key = f"metric:{metric}:host:{host}"
        
        # Store the metric in Redis
        await redis.rpush(key, value)
        
        # Keep only the recent values
        if await redis.llen(key) > window_size:
            await redis.lpop(key)

async def scrape_container_metrics(monitor_host: str, container_names: List[str], window_size: int, session: aiohttp.ClientSession) -> dict:
    
    try:
        # Make the request
        async with session.get(f"http://{monitor_host}/metrics", params={'container_names' : container_names}) as response:
            response = await response.json()
        
    except:
        # If the request fails, flag all containers as down
        response = {
            container : {
                "cpu_absolute_usage": -1,
                "cpu_percent_usage": -1,
                "memory_absolute_usage": -1,
                "memory_percent_usage": -1,
                "network_input": -1,
                "network_output": -1,
                "disk_read": -1,
                "disk_write": -1
            } for container in container_names
        }
    
    # Execute the write tasks concurrently
    write_tasks = [write_to_redis(redis,
                                  metric,
                                  monitor_host,
                                  container_name,
                                  response[container_name][metric],
                                  window_size
                                ) for container_name, metric in product(
                                    container_names, ['cpu_absolute_usage', 
                                                      'cpu_percent_usage', 
                                                      'memory_absolute_usage', 
                                                      'memory_percent_usage', 
                                                      'network_input', 
                                                      'network_output', 
                                                      'disk_read', 
                                                      'disk_write'])
                ]
    await asyncio.gather(*write_tasks)

async def get_network_latency(monitor_host: str, window_size: int) -> float:
    
    # Get the latency
    latencies = []
    for _ in range(5):
        try:
            # Measure the latency
            latency = ping(monitor_host.split(':')[0], timeout=0.1, unit='ms')
            latencies.append(latency)
        except Exception as e:
            # If the request fails, append None
            latencies.append(None)
    
    # If any request fails, set the latency to -1
    if any(latency == None for latency in latencies):
        latency = -1
    else:
        # Calculate the average latency
        latency = sum(latencies) / len(latencies)
    
    # Store the latency in Redis
    await write_to_redis(redis,
                          "network_latency",
                          monitor_host,
                          None,
                          latency,
                          window_size
                        )

async def scraper_loop(monitor_dicts: Dict[str,List[str]], interval: int, window_size: int) -> None:
    
    # Create a session
    async with aiohttp.ClientSession() as session:
        
        # Loop forever
        while True:
            
            # Get the current time
            start_time = perf_counter()
            
            # Fetch metrics for each endpoint
            scrape_tasks = [scrape_container_metrics(monitor_host, container_names, window_size, session) for monitor_host, container_names in monitor_dicts.items()]
            scrape_tasks.extend([get_network_latency(monitor_host, window_size) for monitor_host in monitor_dicts.keys()])
            await asyncio.gather(*scrape_tasks)
            
            # Publish redis message
            await redis.publish("dashboard_metrics", "uploaded")
            
            # Wait for the next interval
            remaining_time = interval - (perf_counter() - start_time)
            if remaining_time > 0:
                await asyncio.sleep(remaining_time)

# Connect to Redis
redis = AsyncRedis.from_url(url=f"redis://{os.environ.get("HOST")}:{os.environ.get("DB_PORT")}", decode_responses=True)

# Clear the Redis database
redis.flushdb()

# Run the scraper
asyncio.run(
    scraper_loop(
        json.loads(os.environ.get("ENDPOINTS")),
        int(os.environ.get("INTERVAL")),
        int(os.environ.get("WINDOW_SIZE"))
    )
)

File: containers/scraper/go/scraper.go
package main

import (
	"context"
	"fmt"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	"encoding/json"
	"net/http"
	"net/url"

	probing "github.com/prometheus-community/pro-bing"
	"github.com/redis/go-redis/v9"
)

var ctx = context.Background()

func writeToRedis(redisClient *redis.Client, metric string, host string, containerName string, value string, windowSize int, wg *sync.WaitGroup) {
	// Close the wait group if it is not nil
	if wg != nil {
		defer wg.Done()
	}

	// Remove port from host if present
	if strings.Contains(host, ":") {
		host = strings.Split(host, ":")[0]
	}

	if containerName != "" {
		// Get key
		key := fmt.Sprintf("metric:%s:host:%s:container:%s", metric, host, containerName)

		// Write to Redis
		redisClient.RPush(ctx, key, value)

		if redisClient.LLen(ctx, key).Val() > int64(windowSize) {
			// Remove the oldest element
			redisClient.LPop(ctx, key)
		}
	} else {
		// Get key
		key := fmt.Sprintf("metric:%s:host:%s", metric, host)

		// Write to Redis
		redisClient.RPush(ctx, key, value)

		if redisClient.LLen(ctx, key).Val() > int64(windowSize) {
			// Remove the oldest element
			redisClient.LPop(ctx, key)
		}
	}
}

func scrapeContainerMetrics(monitorHost string, containerNames []string, windowSize int, redisClient *redis.Client, overallWG *sync.WaitGroup) {
	// Close the wait group if it is not nil
	if overallWG != nil {
		defer overallWG.Done()
	}

	// Get the base URL
	baseURL := fmt.Sprintf("http://%s/metrics", monitorHost)

	// Store the container names in a parameter
	params := url.Values{}
	for _, containerName := range containerNames {
		params.Add("container_names", containerName)
	}

	// Create the URL with parameters
	fullURL := fmt.Sprintf("%s?%s", baseURL, params.Encode())

	// Make the HTTP GET request
	response, err := http.Get(fullURL)

	// Make sure the request was successful
	metrics := make(map[string]map[string]float64)
	if err != nil {
		fmt.Println("Error scraping container metrics:", err)
		// If there is an error, set the metrics to -1
		for _, containerName := range containerNames {
			metrics[containerName] = map[string]float64{
				"cpu_absolute_usage":    -1,
				"cpu_percent_usage":     -1,
				"memory_absolute_usage": -1,
				"memory_percent_usage":  -1,
				"network_input":         -1,
				"network_output":        -1,
				"disk_read":             -1,
				"disk_write":            -1,
			}
		}
	} else {
		// Decode the JSON response
		json.NewDecoder(response.Body).Decode(&metrics)
	}

	// Close the response body
	defer response.Body.Close()

	// Set up a wait group
	var wg sync.WaitGroup

	// Write the metrics to Redis
	for containerName, containerMetrics := range metrics {
		for metric, value := range containerMetrics {
			wg.Add(1)
			go writeToRedis(redisClient, metric, monitorHost, containerName, fmt.Sprintf("%f", value), windowSize, &wg)
		}
	}

	// Wait for all goroutines to finish
	wg.Wait()
}

func getNetworkLatency(monitorHost string, windowSize int, redisClient *redis.Client, overallWG *sync.WaitGroup) {
	// Close the wait group if it is not nil
	if overallWG != nil {
		defer overallWG.Done()
	}

	// Initialize the ping client
	pinger, err := probing.NewPinger(strings.Split(monitorHost, ":")[0])
	if err != nil {
		fmt.Println("Error initializing ping client:", err)
	}

	// Set the number of packets to send
	pinger.Count = 5
	pinger.Timeout = time.Second / 10

	// Send packets
	err = pinger.Run()
	var averageTime string
	if err != nil {
		fmt.Println("Error sending packets:", err)
		averageTime = "-1"

	} else {
		// Get the average time
		averageTime = fmt.Sprintf("%f", float64(pinger.Statistics().AvgRtt)/float64(time.Millisecond))
	}

	// Write the latency to Redis
	writeToRedis(redisClient, "network_latency", monitorHost, "", averageTime, windowSize, nil)
}

func main() {
	// Connect to Redis
	Address := fmt.Sprintf("%s:%s", os.Getenv("HOST"), os.Getenv("DB_PORT"))
	redisClient := redis.NewClient(&redis.Options{
		Addr: Address,
	})
	defer redisClient.Close()

	// Clear the Redis database
	err := redisClient.FlushDB(ctx).Err()
	if err != nil {
		fmt.Println("Error flushing Redis database:", err)
		return
	}

	// Get the hosts and endpoints
	rawMonitorDict := os.Getenv("ENDPOINTS")
	var monitorDict map[string][]string
	err = json.Unmarshal([]byte(rawMonitorDict), &monitorDict)
	if err != nil {
		fmt.Println("Error parsing ENDPOINTS:", err)
		return
	}

	// Get interval
	rawInterval, err := strconv.Atoi(os.Getenv("INTERVAL"))
	interval := time.Duration(rawInterval) * time.Second
	if err != nil {
		fmt.Println("Error parsing INTERVAL:", err)
		return
	}

	// Get window size
	windowSize, err := strconv.Atoi(os.Getenv("WINDOW_SIZE"))
	if err != nil {
		fmt.Println("Error parsing WINDOW_SIZE:", err)
		return
	}

	// Define waitgroup
	var wg sync.WaitGroup

	for {
		// Get current time
		start := time.Now()

		// Scrape metrics for each host
		for host, containerNames := range monitorDict {
			wg.Add(2)
			go scrapeContainerMetrics(host, containerNames, windowSize, redisClient, &wg)
			go getNetworkLatency(host, windowSize, redisClient, &wg)
		}

		// Wait for all goroutines to finish
		wg.Wait()

		// Publish redis message
		err := redisClient.Publish(ctx, "dashboard_metrics", "uploaded").Err()
		if err != nil {
			fmt.Println("Error publishing message to Redis:", err)
		}

		// Get the elapsed time
		elapsed := time.Since(start)

		// Wait for the remaining time
		if elapsed < interval {
			time.Sleep(time.Second*10 - elapsed)
		}
	}
}

// To compile:
// CGO_ENABLED=0 go build -o monitor
